### PR TITLE
[TTAHUB-3026] Self Service Automation - Database Backup - backup retention

### PR DESCRIPTION
**Pull request by GarrettEHill**
## Description of change
Implementation of the following as part of the backup process:

- All backups for the 0-30 days (This includes daily backups, deployment triggered backups, manually triggered backups).
- Daily backups for 31-60 days
- Monday daily backups + 1st & 15th of the month backups for for 61-90 days
- 1st of the month for backups older then 90 days

New features to the latest_backup.sh script
- flag based options
- listing of available files
- requesting a specific file
- download and verify
- erase file from server

## How to test

- run "manual-trigger" on the pipeline to trigger backup and backup retention rules be applied
- use latest_backup.sh script to verify all remaining files follow the retention schedule

## Issue(s)

* https://jira.acf.gov/browse/TTAHUB-3026

### Every PR

- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Before merge to main

- [ ] OHS demo complete
- [ ] Ready to create production PR

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: bin/latest_backup.sh [hunk 1]
```diff
@@ -117,18 +136,180 @@ generate_presigned_urls() {
...
+            echo "Error: File s3://${bucket_name}/${file} does not exist."
+            exit 1
+        fi
     done
+
     echo "${urls[@]}"
 }
 
+# Function to list all ZIP files in the same S3 path as the latest backup
+list_all_zip_files() {
```
Comment (thewatermethod) [inline bin/latest_backup.sh:155]:
```sh
sh bin/latest_backup.sh --list-zip-files
```

Screenshot 2024-07-30 at 9 15 12 AM

Comment (GarrettEHill) [reply bin/latest_backup.sh:155]: I dont see that issue? can you try again

Comment (GarrettEHill) [reply bin/latest_backup.sh:155]: I am unable to reproduce this error

**Review by nvms — COMMENTED** [review]
```bash
Head-Start-TTADP sh bin/latest_backup.sh --list-zip-files
Current cf version (7.5.0+0ad1d63.2022-06-04) is less than 8.0.0. Please update your cf CLI.
Using service key name: ttahub-db-backups-key-962B0A02-A132-4B62-85AB-1F95ECC1CF35 for service instance: ttahub-db-backups
Checking if service key ttahub-db-backups-key-962B0A02-A132-4B62-85AB-1F95ECC1CF35 exists...
Service key ttahub-db-backups-key-962B0A02-A132-4B62-85AB-1F95ECC1CF35 does not exist.
Creating service key with name ttahub-db-backups-key-962B0A02-A132-4B62-85AB-1F95ECC1CF35...
Creating service key ttahub-db-backups-key-962B0A02-A132-4B62-85AB-1F95ECC1CF35 for service instance ttahub-db-backups as jon.pyers@adhocteam.us...
OK

Checking if service key ttahub-db-backups-key-962B0A02-A132-4B62-85AB-1F95ECC1CF35 exists...
Service key ttahub-db-backups-key-962B0A02-A132-4B62-85AB-1F95ECC1CF35 exists.
Verifying AWS credentials...

Could not connect to the endpoint URL: "https://sts.null.amazonaws.com/"
Failed to verify AWS credentials. Attempt 1 of 5.

Could not connect to the endpoint URL: "https://sts.null.amazonaws.com/"
Failed to verify AWS credentials. Attempt 2 of 5.

Could not connect to the endpoint URL: "https://sts.null.amazonaws.com/"
Failed to verify AWS credentials. Attempt 3 of 5.

Could not connect to the endpoint URL: "https://sts.null.amazonaws.com/"
Failed to verify AWS credentials. Attempt 4 of 5.

Could not connect to the endpoint URL: "https://sts.null.amazonaws.com/"
Failed to verify AWS credentials. Attempt 5 of 5.
```

**Comment by thewatermethod** [comment]
as per your request, @GarrettEHill 

```sh
> $ cf target                                                                                                                                        ⬡ 18.20.3 [±TTAHUB-3026/backup-retention ✓]
API endpoint:   https://api.fr.cloud.gov
API version:    3.172.0
user:           matt.bevilacqua@adhocteam.us
org:            hhs-acf-ohs-tta
space:          ttahub-prod
                                                                                                                                                                                                 
matthewbevilacqua@matthewbevilacqua-C02FL2B7MD6P ~/Head-Start-TTADP                                                                                                                   [11:46:07]
> $ sh bin/latest_backup.sh --list-zip-files                                                                                                         ⬡ 18.20.3 [±TTAHUB-3026/backup-retention ✓]
Current cf version (8.7.10+5b7ce3c.2024-04-04) is greater than or equal to 8.0.0.
Using service key name: ttahub-db-backups-key-7199C716-9093-48A8-B6AA-BD7B73B5A8DA for service instance: ttahub-db-backups
Checking if service key ttahub-db-backups-key-7199C716-9093-48A8-B6AA-BD7B73B5A8DA exists...
Service key ttahub-db-backups-key-7199C716-9093-48A8-B6AA-BD7B73B5A8DA does not exist.
Creating service key with name ttahub-db-backups-key-7199C716-9093-48A8-B6AA-BD7B73B5A8DA...
Creating service key ttahub-db-backups-key-7199C716-9093-48A8-B6AA-BD7B73B5A8DA for service instance ttahub-db-backups as matt.bevilacqua@adhocteam.us...
OK

Checking if service key ttahub-db-backups-key-7199C716-9093-48A8-B6AA-BD7B73B5A8DA exists...
Service key ttahub-db-backups-key-7199C716-9093-48A8-B6AA-BD7B73B5A8DA exists.
Verifying AWS credentials...
AWS credentials are valid.
ZIP files in S3 bucket:
Name                                               pwd   md5   sha256 size(zip)       age(days)
awk: syntax error at source line 17
 context is
                    >>>  files[base][ <<< ext] = 1
awk: illegal statement at source line 17
awk: illegal statement at source line 17
Service key deletion is not enabled. Set DELETION_ALLOWED=yes to enable.

**Comment by thewatermethod** [comment]
```sh
sh bin/latest_backup.sh -d  
```

Looks like Macs don't have wget installed by default?

```sh
download: s3://cg-7409cf93-8401-40e3-b73a-f5731446b79a/production/production-latest-backup.txt to ../../../tmp/latest-backup.txt
bin/latest_backup.sh: line 212: wget: command not found
Verifying SHA-256 checksum...
sha256sum: production-2024-07-31-14-38-55-UTC.sql.zip: No such file or directory
production-2024-07-31-14-38-55-UTC.sql.zip: FAILED open or read
sha256sum: WARNING: 1 listed file could not be read
SHA-256 checksum verification failed.
```

**Comment by thewatermethod** [comment]
Also, running 

```sh
sh bin/latest_backup.sh -l
```
immediately after the above seems to reference a token with the wrong permissions?

```sh
An error occurred (AccessDenied) when calling the ListObjectsV2 operation: Access Denied
No ZIP files found in S3 bucket.
Service key deletion is not enabled. Set DELETION_ALLOWED=yes to enable.

**Review by thewatermethod — APPROVED** [review]

**Pull request merged**